assets/src/models/ladderVehiclePosition.ts

```diff
@@ -0,0 +1,156 @@
+      ): LadderVehiclePosition[] => {
+        const overlappers = overlappingPreviousVehicles(
+          previousLadderVehiclePositions,
+          ladderVehiclePosition
+        )
+        const occupiedLanes: number[] = overlappers.reduce(
+          (lanes: number[], overlapper: LadderVehiclePosition) =>
+            overlapper.lane !== undefined ? [...lanes, overlapper.lane] : lanes,
+          []
+        )
+        const lane = firstOpenLane(occupiedLanes)
+
+        return [
+          {
+            ...ladderVehiclePosition,
+            lane,
+          },
+          ...previousLadderVehiclePositions,
+        ]
+      },
+      []
+    )
+
+export const xInLaneWithOffset = (ladderVehicleHorizontalOffset: number) => ({
+  x,
+  vehicleDirection,
+  lane,
+}: LadderVehiclePosition): number => {
+  const laneOffset = (lane || 0) * ladderVehicleHorizontalOffset
+  return vehicleDirection === VehicleDirection.Up
+    ? x + laneOffset
+    : x - laneOffset
+}
+
+export const byDirectionAndY = (
+  a: LadderVehiclePosition,
+  b: LadderVehiclePosition
+): number => {
+  if (a.vehicleDirection === VehicleDirection.Down) {
+    if (b.vehicleDirection === VehicleDirection.Up) {
+      return -1
+    } else {
+      if (a.y > b.y) {
+        return -1
+      } else if (b.y > a.y) {
+        return 1
+      } else {
+        return 0
+      }
+    }
+  } else {
+    if (b.vehicleDirection === VehicleDirection.Down) {
+      return 1
+    } else {
+      if (a.y < b.y) {
+        return -1
+      } else if (b.y < a.y) {
+        return 1
+      } else {
+        return 0
+      }
+    }
+  }
+}
+
+const overlappingPreviousVehicles = (
+  previousLadderVehiclePositions: LadderVehiclePosition[],
+  ladderVehiclePosition: LadderVehiclePosition
+): LadderVehiclePosition[] =>
+  previousLadderVehiclePositions.filter(previous =>
+    areOverlapping(previous, ladderVehiclePosition)
+  )
+
+export const areOverlapping = (
+  a: LadderVehiclePosition,
+  b: LadderVehiclePosition
+): boolean => {
+  return (
+    a.vehicleDirection === b.vehicleDirection &&
+    areOverlappingRanges(yRange(a), yRange(b))
+  )
+}
+
+const areOverlappingRanges = (
+  [minA, maxA]: Range,
+  [minB, maxB]: Range
+): boolean => minA < maxB && minB < maxA
+
+const yRange = ({ y, vehicleDirection }: LadderVehiclePosition): Range => {
+  if (vehicleDirection === VehicleDirection.Down) {
+    const topY = y + centerOfVehicleGroupY
+    return [topY - heightOfVehicleGroup, topY]
+  } else {
+    const topY = y - centerOfVehicleGroupY
+    return [topY, topY + heightOfVehicleGroup]
+  }
+}
+
+export const firstOpenLane = (occupiedLanes: number[] | undefined): number => {
+  if (!occupiedLanes || !occupiedLanes.length) {
+    return 0
+  }
+
+  const gappedLanes: number[] = range(occupiedLanes.length).filter(
+    lane => !occupiedLanes.includes(lane)
+  )
+  if (gappedLanes && gappedLanes.length) {
+    return gappedLanes[0]
+  }
+
+  return Math.max(...occupiedLanes) + 1
```

Comment:
Could there be a simpler way to implement this by sorting the input and walking through it til you find the first gap?

Reply:
That was actually my first implementation. It ended up being much less simple because their might not be a gap, or might not be any occuied lanes. This is much more straightforward and these list sizes are never going to be that large.

Reply:
What about:
```
const sortedOccupiedLanes = occupiedLanes.slice().sort()
for (var i = 0; i < sortedOccupiedLanes.length; i++) { 
  if (sortedOccupiedLanes[i] !== i) {
    return i
  }
}
return sortedOccupiedLanes.length
```
It's not very functional, but I think it's still simpler, and should cover if the list is empty (do the loop 0 times, then return 0), and full (do the loop n times, then return n).

Reply:
That works.